File: 316-remove-duplicate-letters/remove-duplicate-letters.java
import java.util.*;

public class Solution {
    public String removeDuplicateLetters(String s) {
        int[] put = new int[26];
        int[] fre = new int[26];

        for (int i = 0; i < s.length(); i++) {
            fre[s.charAt(i) - 'a']++;
        }

        Stack<Character> st = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            fre[ch - 'a']--;

            if (put[ch - 'a'] == 1) {
                continue;
            }

            while (!st.isEmpty() && st.peek() > ch && fre[st.peek() - 'a'] > 0) {
                put[st.peek() - 'a'] = 0;
                st.pop();
            }

            st.push(ch);
            put[ch - 'a'] = 1;
        }

        // Build result from stack
        StringBuilder ans = new StringBuilder();
        for (char c : st) {
            ans.append(c);
        }

        return ans.toString();
    }
}

File: 316-remove-duplicate-letters/remove-duplicate-letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<int>put(26,0);
        vector<int>fre(26,0);
        for(int i=0;i<s.length();i++)
        {
                fre[s[i]-'a']++;
        }
        stack<char>st;
        for(int i=0;i<s.length();i++)
        {
                if(st.empty())
                {
                     
                     st.push(s[i]);
                     put[s[i]-'a']=1;
                }
                else{
                       if(put[s[i]-'a']==1)
                       {
                           fre[s[i]-'a']--;
                           continue;
                       }
                        while(!st.empty()&&(st.top()-'a')>=(s[i]-'a')&&fre[st.top()-'a']>1)
                        {
                                    
                                        fre[st.top()-'a']--;
                                         put[st.top()-'a']=0;
                                        st.pop();
                        }
                         if( put[s[i]-'a']==0){
                        st.push(s[i]);
                        put[s[i]-'a']=1;
                         }

                }
        }
        string ans;
        while(!st.empty())
        {
              ans+=st.top();
              st.pop();
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};